lib/events/filesessions/fileasync_chaos_test.go

```diff
@@ -0,0 +1,229 @@
+// +build !race
+
+/*
+Copyright 2020 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+*/
+
+package filesessions
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/teleport/lib/session"
+	"github.com/gravitational/teleport/lib/utils"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	log "github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
+	"go.uber.org/atomic"
+)
+
+// TestUploadChaos introduces failures in all stages of the async
+// upload process and verifies that the system is working correctly.
+//
+// Data race detector slows down the test significantly (10x+),
+// that is why the test is skipped when tests are running with
+// `go test -race` flag or `go test -short` flag
```

Comment:
We run integration test with race detector as well. I think we should have a separate class of tests we run without race detector.

I would like to have tests that run for a while and try to break the system in different ways.

Reply:
@webvictim can we add a line to run `go test -test.run=TestChaos ./lib/....` to activate those tests?

Reply:
I stand corrected, integration tests are running without `-race` flag, however I think it will be good to have `TestChaos` step that is local, but does not activate `-race` detector.

Reply:
@webvictim @awly I've added another build target `make chaos` and `make -C build.assets chaos` alongside extra drone step. What do you folks think? Take a look and let me know please.

Reply:
For others' reference, `TestChaosUpload` takes:
- 0.3s without race detector
- 12.4s with race detector
12s is slow enough to be annoying, although we have much slower unit tests.

I'd prefer to either add it to `make test` without race detector or `make integration` with (eventually).

If we're going to add a new category of tests, it could be based on how they are used:
- `make test` - "fast" tests for local dev iteration
- `make integration` - "slow" tests for each PR running in CI
- `make stress` - "infinite" tests that hammer at the code for a fixed period of time, running as a cronjob
  - fuzzers also fit here

P.S. integration tests don't run with the race detector because we didn't get to fixing all the races yet. They should eventually run with the race detector.

Reply:
The biggest problem with adding this as a completely separate step to Drone is the fact that the tests take 0.3 seconds to run but the setup (downloading the buildbox image, starting the Docker container etc) takes 22 seconds. Given that our unit test runs are already taking ~4-5 minutes and integration tests ~4-5 minutes, I don't think it's a huge problem to add them to one of the other runs.

I agree with @awly's idea of saying that ultimately, all unit tests should be 'quick' tests you can run locally for iterative development. Anything that takes a long time should be part of integration.